webapp/src/js/services/search.js

```diff
@@ -52,39 +52,62 @@ var _ = require('underscore'),
+        var interval = CalendarInterval.getCurrent(settings.monthStartDate),
+            visitStats = {};
 
-        return DB().query('medic-client/contacts_by_last_visited', {
-          reduce: false,
-          keys: searchResults
-        }).then(function(result) {
-          var visitStats = {},
-              interval = CalendarInterval.getCurrent(settings.monthStartDate);
-
-          result.rows.forEach(function(row) {
-            var stats = visitStats[row.key] || { lastVisitedDate: -1, visitDates: [] };
+        searchResults.forEach(function(id) {
+          visitStats[id] = { lastVisitedDate: -1, visitDates: [] };
+        });
 
-            stats.lastVisitedDate = Math.max(stats.lastVisitedDate, row.value);
+        var getVisitsInInterval = function() {
+          return DB()
+            .query('medic-client/visits_by_date', { start_key: interval.start, end_key: interval.end })
+            .then(function(result) {
+              result.rows.forEach(function (row) {
+                if (visitStats[row.value]) {
+                  visitStats[row.value].visitDates.push(moment(row.key).startOf('day').valueOf());
+                }
+              });
+            });
+        };
 
-            if (row.value >= interval.start && row.value <= interval.end) {
-              stats.visitDates.push(moment(row.value).startOf('day').valueOf());
+        var setLastVisitedDate = function(rows) {
+          rows.forEach(function(row) {
+            if (visitStats[row.key]) {
+              visitStats[row.key].lastVisitedDate = row.value && row.value.max || row.value;
             }
-
-            visitStats[row.key] = stats;
           });
-
-          return Object.keys(visitStats).map(function(key) {
-            return {
-              key: key,
-              value: {
-                lastVisitedDate: visitStats[key].lastVisitedDate,
-                visitCount: _.uniq(visitStats[key].visitDates).length,
-                visitCountGoal: settings.visitCountGoal
-              }
-            };
+        };
+
+        var getLastVisited = function() {
+          if (extendedResults) {
+            // when sorting by last visited date, we receive the data from Search library
+            return setLastVisitedDate(extendedResults);
+          }
+
+          return DB()
+            .query('medic-client/contacts_by_last_visited', { reduce: true, group: true })
+            .then(function(result) {
+              setLastVisitedDate(result.rows);
+            });
+        };
+
+        return getVisitsInInterval()
+          .then(getLastVisited)
+          .then(function() {
+            return Object.keys(visitStats).map(function(key) {
+              return {
+                key: key,
+                value: {
+                  lastVisitedDate: visitStats[key].lastVisitedDate,
+                  visitCount: _.uniq(visitStats[key].visitDates).length,
+                  visitCountGoal: settings.visitCountGoal
```

Comment:
I know this is not a new problem, but it seems unnecessary to me that we include this settings value in every item. No concrete feedback for this change.

Reply:
Agreed - it would be nicer to set this once only. It doesn't need to block this PR however...